index.js

```diff
@@ -1,212 +1,236 @@
-    }
-  , getDigest: function () {
-      return this._digest
-    }
-  , handlePutResponse: function (error, resp) {
-      if (error) {
-        console.error('ERROR', error)
-        this.options.error && this.options.error(error)
-        this.abort(error)
-      }
-      else {
-        this._inProgress = false
-        this.start()
-      }
-    }
-  , _mergeHeaders: function (file) {
-      var o = {}
-      if (file.match('.gz')) {
-        o['ContentEncoding'] = 'gzip'
-        o['CacheControl'] = 'max-age=1314000'
+      })
+    })
+    .catch(err => {
+      if (/already exists/i.test(err.message)) {
+        done()
+      } else {
+        done(err)
       }
-      o['ContentType'] = mime.lookup(file)
-      // overwrite `gz` headers
-      if (file.match('.js')) o['ContentType'] = 'application/javascript'
-      if (file.match('.css')) o['ContentType'] = 'text/css'
-      return o
-    }
-  , _addToDigest: function (file) {
-      var md5File = this.prefix + this.hashContents(file)
-        , s3FileName = file.substring(this.path.length)
+    })
+  }
 
-      this._digest[s3FileName] = md5File
-    }
-  })
+  s3FilePreCheck(file) {
+    // check for existence of file and reject if it exists to avoid uploading same file again
+    return new Promise((resolve, reject) => {
+      this.client.getObject({
+        Bucket: this.bucket,
+        Key: file
+      }, err => {
+        if (this.httpResponse.statusCode == 404) {
+          resolve()
+        } else if (this.httpResponse.statusCode == 200) {
+          reject(new Error('File already exists - ' + file))
+        } else {
+          reject(
+            new Error('S3 File precheck failed [' + this.httpResponse.statusCode + '] for ' +
+              file + ': ' + err.message)
+          )
+        }
+      })
+    })
+  }
+
+  abort(e) {
+    this._timer && clearTimeout(this._timer)
+    debug('S3 sync aborted', e)
+    this.options.complete && this.options.complete(new Error('Sync aborted'))
+  }
+
+  getDigest() {
+    return this._digest
+  }
+
+  handlePutResponse(error) {
+    if (error) {
+      this.options.error && this.options.error(error)
+      this.abort(error)
+    } else {
+      this._inProgress = false
+      this.start()
+    }
+  }
+}
+
+module.exports = {
+  S3Sync
+}
+
+function mergeHeaders(file) {
+  const o = {}
+  if (file.match('.gz')) {
+    o['ContentEncoding'] = 'gzip'
+    o['CacheControl'] = 'max-age=1314000'
+  }
+  o['ContentType'] = mime.lookup(file)
+  // overwrite `gz` headers
+  if (file.match('.js')) {
+    o['ContentType'] = 'application/javascript'
+  }
+  if (file.match('.css')) {
+    o['ContentType'] = 'text/css'
+  }
+  return o
+}
 
-module.exports.s3sync = function (config, options) {
-  return new S3Sync(config, options)
+function addToDigest(file) {
+  const md5File = this.prefix + this.hashContents(file)
+  const s3FileName = file.substring(this.path.length)
+  this._digest[s3FileName] = md5File
```

Comment:
Since `addToDigest` relies on `this` context to be an `S3Sync` instance, it would make more sense for it to be a class method.

Reply:
This and `mergeHeaders` were "private" methods.
TBH - I think the class interface has way too many exposed methods so this was an attempt to start moving private methods to truly be private.
I was hoping we could follow up by moving more functions out of the class and only leaving what truly needs to be called from implementations exposed.

Reply:
Heads-up, babel now offers a syntax to make class methods private: https://babeljs.io/blog/2018/12/03/7.2.0.html#private-instance-methods-8654-https-githubcom-babel-babel-pull-8654

It is on-track to become standard (currently TC39 Stage 3): https://github.com/tc39/proposal-private-methods

Reply:
`#` as a prefix is weird
#prefixforprivateisweird